### PR TITLE
Added Bson-Kotlin Array Codec

### PR DESCRIPTION
**Pull request by rozza**
Adds Kotlin array support to the bson-kotlin library

JAVA-5122

### CODE REVIEW

File: bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                    Pair(kClass.java.componentType.kotlin.javaObjectType as Class<Any>, emptyList())
+                } else {
+                    when (val pType = typeArguments[0]) {
+                        is Class<*> -> Pair(pType as Class<Any>, emptyList())
+                        is ParameterizedType -> Pair(pType.rawType as Class<Any>, pType.actualTypeArguments.toList())
+                        else -> Pair(Object::class.java as Class<Any>, emptyList())
+                    }
+                }
+            val codec =
+                if (nestedTypes.isEmpty()) codecRegistry.get(valueClass) else codecRegistry.get(valueClass, nestedTypes)
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt:50]: Yes, we follow the main kotlin coding standards. 
See: https://kotlinlang.org/docs/control-flow.html#if-expression

It looks unusual as in Java we'd use a tenary operator but in Kotlin there is no such thing.

File: bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        reader.readEndArray()
+        return unboxed(data)
+    }
+
+    fun boxed(arrayValue: R): Iterator<V?> {
+        val boxedValue =
+            if (!isPrimitiveArray) {
+                (arrayValue as Array<V?>).iterator()
+            } else if (arrayValue is BooleanArray) {
+                arrayValue.toList().iterator()
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt]: It was used to stop a spotbugs error, however, using `asIterable` works without triggering the error.

File: bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            val (valueClass, nestedTypes) =
+                if (typeArguments.isEmpty()) {
+                    Pair(kClass.java.componentType.kotlin.javaObjectType as Class<Any>, emptyList())
+                } else {
+                    when (val pType = typeArguments[0]) {
+                        is Class<*> -> Pair(pType as Class<Any>, emptyList())
+                        is ParameterizedType -> Pair(pType.rawType as Class<Any>, pType.actualTypeArguments.toList())
+                        else -> Pair(Object::class.java as Class<Any>, emptyList())
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson-kotlin/src/main/kotlin/org/bson/codecs/kotlin/ArrayCodec.kt]: As its a single block, I opted to add a comment instead.

**Comment by rozza** [comment]
Thanks @vbabanin the when block looks much better

**Review by vbabanin — APPROVED** [review]

**Pull request merged**